cvastrophoto/rops/denoise/diffusion.py

```diff
@@ -0,0 +1,160 @@
+# cython: infer_types=True
+
+import PIL
+import math
+import numpy
+import cython
+import logging
+from skimage import color, measure
+from ..base import PerChannelRop
+if not cython.compiled:
+    globals().update(dict(fabs=abs))
+
+
+logger = logging.getLogger(__name__)
+
+class DiffusionRop(PerChannelRop):
+    def __init__(self,raw=None,T=0.0025,L=1.0/180,R=2,thr=3,steps=20,dt=1.0/7):
+        super(DiffusionRop,self).__init__(raw)
+        self.T = T
+        self.L = L
+        self.R = R
+        self.thr = thr
+        self.dt = dt
+        self.steps = steps
+    def process_channel(self,data,detected=None): 
+        accelerated=False
+        denoiser = Denoise(self.L,accelerated)
+        return denoiser.modifiedPMdiffusion(data,self.steps,self.R,self.T,self.thr,self.dt)
```

Comment:
If `data` isn't of type `numpy.float32`, this code will break.

So you should forcibly cast it to `float32` to make sure:

``` python
if data.dtype.char != 'f':
    data = data.astype(numpy.float32)
```